### DISCUSSION TITLE
Update the Version property of csproj which has updated dependencies

### DISCUSSION THREAD

**Discussion by bjuraga** [renovatebot/renovate#17780] [Request Help]
Hi, i have been trying to find a way to do update the nuget packages Version property in the csproj file when Renovate updates a dependency.

Our setup:
We use Renovate self-hosted in Azure DevOps, to renovate nuget, npm and docker dependencies. 
We have a monorepo for all nuget packages, they use the SDK style csproj files.
Each csproj has a Version property that we manually increased before we started using renovate, we are not mature ATM to use auto versioning.

What i tried:
Set the bumpVersion for nuget, did not do anything:
config.js
```hostRules: [
    {
      hostType:"nuget",
	  ...
      bumpVersion: "minor"
    }
  ]
```

Now trying to use postUpgradeTasks, my idea is to run a powershell script, but it never seems to run, i dont see any logs about it running or not even with trace level logging. 
Configuration:
config.js
```
  allowPostUpgradeCommandTemplating: true,
  allowedPostUpgradeCommands: ['powershell -noexit .\bump-version.ps1']
```
renovate.json
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "postUpgradeTasks": {
    "commands": [
        "powershell -noexit .\bump-version.ps1"
    ],
    "fileFilters": ["**.csproj"],
    "executionMode": "branch"
  }
}
```
And the contents of the powershell is 
```
@(git status) | Where-Object{ "$_".Contains(".csproj") } | ForEach-Object { 
    $fileName = "$PSScriptRoot\$("$_".Replace("modified:", '').Trim())";
    $xml = [xml](Get-Content $fileName)
    $xml.Project.PropertyGroup.Version = "1.2.7"; 
    $xml.Save($fileName);
     }
```
The powershell relies on the csprojs being changed by Renovate in case a dependency is updated at the moment the postUpgradeTask runs. 
It then gets all changed csproj files (from git's point of view) and for each csproj it sets the version. It is hard coded just for debug reasons, but proper semver handling will be added.

I cant find a single case where someone was stuck in the same position as me, most likely everyone does autoversioning?

What would be the way to achieve this? I am hoping there is a native support in Renovate, so i dont need to engineer the second approach, but i dont mind doing the work, just not sure how to set it all up.

**Comment by rarkins** [comment]
https://github.com/renovatebot/renovate/issues/14348

**Reply by bjuraga** [reply]
Thanks @rarkins, i did see that and that gave me the idea to try the first approach with bumpVersion. However, that issue seems stale and not implemented. Hence, my second, more manual attempt. Still hoping my first attempt is just simply misconfigured 🤞

If you could please point me to a good example for bumpVersion implementation, can be any other language/platform/plugin/etc, i can try creating the missing nuget implementation (if missing).

Otherwise, i would have to get postUpgradeTasks to work. I am also open to any suggestions.

**Reply by rarkins** [reply]
I think best if you look at the non-npm implementations in https://github.com/renovatebot/renovate/search?l=TypeScript&q=bumpversion (maven, sbt, helmv3)

**Reply by bjuraga** [reply]
Here's the implementation of bumpVersion for nuget: 
https://github.com/renovatebot/renovate/compare/main...bjuraga:renovate_add_bumpVersion_to_nuget:add_bumpVersion_to_nuget

I 'simply' copied the maven code and tests, made small adjustments and the tests pass, cover 100% the new change and also cover real life scenarios. All credit goes to the maven manager author(s) - thanks!

Can you please take a look when you have time and let me know if this suffices for me creating a PR?

Thanks for pointing me to the correct place @rarkins

**Reply by bjuraga** [reply]
Added docs changes in case the PR makes it in main.

**Reply by rarkins** [reply]
Looks good to me

**Reply by bjuraga** [reply]
https://github.com/renovatebot/renovate/pull/17903

**Reply by bjuraga** [reply]
Seems like somewhere in the plumbing i do not get the current version from the csproj file.

Logs from my POC:
```
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=true (repository=ORG/poc-renovate, branch=renovate/sentry.aspnetcore-3.x)
TRACE: nuget.extractPackageFile() (repository=ORG/poc-renovate, packageFile=poc.csproj, branch=renovate/sentry.aspnetcore-3.x)
TRACE: nuget.extractPackageFile() (repository=ORG/poc-renovate, packageFile=poc.csproj, branch=renovate/sentry.aspnetcore-3.x)
DEBUG: Branch dep is already updated (repository=ORG/poc-renovate, packageFile=poc.csproj, branch=renovate/sentry.aspnetcore-3.x)
       "depName": "Sentry.Aspnetcore"
DEBUG: Checking if we should bump project version (repository=ORG/poc-renovate, branch=renovate/sentry.aspnetcore-3.x)
       "bumpVersion": "minor",
       "currentValue": undefined
 WARN: Unable to bump project version, project has no version (repository=ORG/poc-renovate, branch=renovate/sentry.aspnetcore-3.x)
DEBUG: No content changed (repository=ORG/poc-renovate, packageFile=poc.csproj, branch=renovate/sentry.aspnetcore-3.x)
       "depName": "Sentry.Aspnetcore"
DEBUG: No package files need updating (repository=ORG/poc-renovate, branch=renovate/sentry.aspnetcore-3.x)
``` 
You can see the outcome of the check for `Checking if we should bump project version` getting undefined.
The poc.csproj content:
```
<Project Sdk="Microsoft.NET.Sdk">

    <PropertyGroup>
        <TargetFrameworks>net6.0</TargetFrameworks>
        <PackageId>POC.Renovate</PackageId>
        <Version>3.16.0</Version>
    </PropertyGroup>

    <ItemGroup>
        ........
        <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
    </ItemGroup>
</Project>
```

Any ideas on what might be missing? I would expect the bug is now in [this file](https://github.com/renovatebot/renovate/blob/ff650ce8225abc6e88909d1e3a5f22bee370a2da/lib/workers/repository/update/branch/get-updated.ts)

**Reply by rarkins** [reply]
Perhaps needs to be added to the extract logic? Check other managers for now they do it

**Reply by bjuraga** [reply]
PR to implement the missing file version #17995
